ir/memory.cpp

```diff
@@ -994,11 +1020,21 @@ void Memory::mk_nonlocal_val_axioms(bool skip_consts) {
     Byte byte(*this, non_local_block_val[i].val.load(offset));
     Pointer loadedptr = byte.ptr();
     expr bid = loadedptr.getShortBid();
+
+    unsigned bid_upperbound = numNonlocals() - 1;
+    if (has_fncall && state->isSource())
+      // exclude fncall mem block (src)
+      bid_upperbound = numNonlocals() - 2;
+    expr loadedptr_cond = !loadedptr.isLocal(false) &&
+                          !loadedptr.isNocapture(false) &&
+                          bid.ule(bid_upperbound);
+
+    if (has_fncall && !state->isSource())
+      // exclude fncall mem block (tgt): bid_upperbound is larger than fnbid
+      loadedptr_cond &= bid != expr::mkUInt(get_fncallmem_bid(), bid);
```

Comment:
I couldn't remove this bit because `numNonlocals()` in tgt returns `num_nonlocals` which is larger than `numNonlocals()` in src.
Even if I couldn't totally remove this check, I think invocation of this can be shrunk because `mkNonlocalValAxioms` is encoding axioms of initial memory blocks in src and tgt even if they are equivalent? For example,
```
@glb = ...
define void @src() { .. }
define void @tgt() { .. }
```
The axiom of initial value of `@glb` is encoded in src and tgt twice; I think it is enough to encode src's one only.

Reply:
You are right that we shouldn't call this function from the memory constructor for the tgt. Nice catch!

But this function is used for the encoding of fn calls as well (called from `Memory::setState()`). We should probably move it to `Memory::mkCallState` so it's only called from src & once.

I suggest you revert these changes and we do the proper fix afterwards to avoid making this patch any more complicated.

Reply:
I'll make the patch for removing the redundant constraint as a follow-up.
The diff here has fncall mem bid check only.